rules/ruleset.go

```diff
@@ -683,3 +685,524 @@ type int64Asc []int64
+		rollupRules[i] = &rule
+	}
+
+	return &ruleSet{
+		uuid:               rsj.UUID,
+		version:            rsj.Version,
+		namespace:          []byte(rsj.Namespace),
+		createdAtNanos:     rsj.CreatedAtNanos,
+		lastUpdatedAtNanos: rsj.LastUpdatedAtNanos,
+		tombstoned:         rsj.Tombstoned,
+		cutoverNanos:       rsj.CutoverNanos,
+		mappingRules:       mappingRules,
+		rollupRules:        rollupRules,
+	}
+}
+
+// RuleSetUpdateHelper helps
+type RuleSetUpdateHelper struct {
+	propagationDelay time.Duration
+}
+
+// NewRuleSetUpdateHelper creates a new RuleSetUpdateHelper struct.
+func NewRuleSetUpdateHelper(propagationDelay time.Duration) RuleSetUpdateHelper {
+	return RuleSetUpdateHelper{propagationDelay: propagationDelay}
+}
+
+// GenUpdateMetadata creates a properly initialized UpdateMetadata object.
+func (r RuleSetUpdateHelper) GenUpdateMetadata() UpdateMetadata {
+	updateTime := time.Now().UnixNano()
+	cutoverTime := updateTime + int64(r.propagationDelay)
+	return UpdateMetadata{lastUpdatedAtNanos: updateTime, cutoverNanos: cutoverTime}
+}
+
+// UpdateMetadata contains fields that should be set with each
+// update to the state of a ruleset.
+type UpdateMetadata struct {
+	cutoverNanos       int64
+	lastUpdatedAtNanos int64
+}
+
+// MappingRuleData is the parts of a mapping rule that an end user would care about.
+// This struct is provided to create or make updates to the mapping rule.
+type MappingRuleData struct {
+	UpdateMetadata
+
+	Name     string            `json:"name" validate:"nonzero"`
+	Filters  map[string]string `json:"filters" validate:"nonzero"`
+	Policies []policy.Policy   `json:"policies" validate:"nonzero"`
+}
+
+// MappingRuleUpdate contains a MappingRuleData along with an ID for a mapping rule.
+// The rule with that ID is meant to updated to the given config.
+type MappingRuleUpdate struct {
+	Data MappingRuleData `json:"config" validate:"nonzero"`
+	ID   string          `json:"id" validate:"nonzero"`
+}
+
+// RollupRuleData is the parts of a rollup rule that an end user would care about.
+// This struct is provided to create or make updates to the rollup rule.
+type RollupRuleData struct {
+	UpdateMetadata
+
+	Name    string            `json:"name" validate:"nonzero"`
+	Filters map[string]string `json:"filters" validate:"nonzero"`
+	Targets []RollupTarget    `json:"targets" validate:"nonzero"`
+}
+
+// RollupRuleUpdate is a RollupRuleConfig along with an ID for a rollup rule.
+// The rule with that ID is meant to updated with a given config.
+type RollupRuleUpdate struct {
+	Data RollupRuleData `json:"data" validate:"nonzero"`
+	ID   string         `json:"id" validate:"nonzero"`
+}
+
+// DeleteData contains data necessary to delete a rule.
+type DeleteData struct {
+	UpdateMetadata
+
+	ID string `json:"id" validate:"nonzero"`
+}
+
+func (rs *ruleSet) AddMappingRule(mr MappingRuleData) error {
+	err := rs.validateMappingRuleUpdate(mr)
+	if err != nil {
+		return err
+	}
+
+	m, err := rs.getMappingRuleByName(mr.Name)
```

Comment:
Any reason why we are getting the mapping rule by name here but the rest of the methods get the mapping rule by id?